api/api.go

```diff
@@ -695,6 +696,76 @@ func (b *BlockChainAPI) getBlockNumber(blockNumberOrHash *rpc.BlockNumberOrHash)
 	return 0, fmt.Errorf("invalid arguments; neither block nor hash specified")
 }
 
+// FeeHistory returns transaction base fee per gas and effective priority fee
+// per gas for the requested/supported block range.
+// blockCount: Requested range of blocks. Clients will return less than the
+// requested range if not all blocks are available.
+// lastBlock: Highest block of the requested range.
+// rewardPercentiles: A monotonically increasing list of percentile values.
+// For each block in the requested range, the transactions will be sorted in
+// ascending order by effective tip per gas and the coresponding effective tip
+// for the percentile will be determined, accounting for gas consumed.
+func (b *BlockChainAPI) FeeHistory(
+	ctx context.Context,
+	blockCount math.HexOrDecimal64,
+	lastBlock rpc.BlockNumber,
+	rewardPercentiles []float64,
+) (*FeeHistoryResult, error) {
+	if blockCount > 1024 {
+		return nil, fmt.Errorf("block count has to be between 1 and 1024")
+	}
+
+	lastBlockNumber := uint64(lastBlock)
+	var err error
+	if lastBlock < 0 {
+		// From the special block tags, we only support "latest".
+		lastBlockNumber, err = b.blocks.LatestEVMHeight()
+		if err != nil {
+			return nil, fmt.Errorf("could not fetch latest EVM block number")
+		}
+	}
+
+	var oldestBlock *hexutil.Big
+	baseFees := []*hexutil.Big{}
+	rewards := [][]*hexutil.Big{}
+	gasUsedRatios := []float64{}
+
+	maxCount := uint64(blockCount)
+	if maxCount > lastBlockNumber {
+		maxCount = lastBlockNumber
+	}
+
+	for i := maxCount; i >= uint64(1); i-- {
+		blockHeight := lastBlockNumber - i + 1
+		block, err := b.blocks.GetByHeight(blockHeight)
+		if err != nil {
+			continue
```

Comment:
this should return an error, why ignore it?

Reply:
From the [description](https://docs.infura.io/api/networks/ethereum/json-rpc-methods/eth_feehistory#parameters):

```bash
blockCount: (integer) Number of blocks in the requested range.
Between 1 and 1024 blocks can be requested in a single query.
If blocks in the specified block range are not available, then only the fee history for available blocks is returned.
```

For non-available blocks, the specification mentions to simply ignore them.

Reply:
what, why would block not be available, that is really weird. Is that only in case that you provide a block count that goes out of range that was indexed?

Reply:
Most likely yes.
While debugging this endpoint with Foundry, I found out that it is usually called with:

```javascript
let response = await web3.eth.getFeeHistory(10, 'latest', [20])
```

which fetches the fee history for 10 blocks, up to the latest. Some nodes may not have all the latest blocks indexer yet.
Anyway, I tried to keep the implementation on par with the specification, so that we do not have any discrepancies.